fish-rust/src/kill.rs

```diff
@@ -0,0 +1,101 @@
+//! The killring.
+//!
+//! Works like the killring in emacs and readline. The killring is cut and paste with a memory of
+//! previous cuts.
+
+use cxx::{CxxVector, CxxWString};
+use std::collections::VecDeque;
+use std::pin::Pin;
+use std::sync::Mutex;
+
+use crate::wchar::WString;
+use crate::wchar_ffi::WCharFromFFI;
+
+#[cxx::bridge]
+mod kill_ffi {
+    extern "Rust" {
+        #[cxx_name = "kill_add"]
+        fn kill_add_ffi(new_entry: &CxxWString);
+        #[cxx_name = "kill_replace"]
+        fn kill_replace_ffi(old_entry: &CxxWString, new_entry: &CxxWString);
+        #[cxx_name = "kill_yank_rotate"]
+        fn kill_yank_rotate_ffi(mut out_front: Pin<&mut CxxWString>);
+        #[cxx_name = "kill_yank"]
+        fn kill_yank_ffi(mut out_front: Pin<&mut CxxWString>);
+        fn kill_entries_len() -> usize;
+        #[cxx_name = "kill_entries"]
+        fn kill_entries_ffi(mut out_entries: Pin<&mut CxxVector<CxxWString>>) -> bool;
+    }
+}
+
+static KILL_LIST: once_cell::sync::Lazy<Mutex<VecDeque<WString>>> =
+    once_cell::sync::Lazy::new(|| Mutex::new(VecDeque::new()));
+
+fn kill_add_ffi(new_entry: &CxxWString) {
+    kill_add(new_entry.from_ffi());
+}
+
+/// Add a string to the top of the killring.
+pub fn kill_add(new_entry: WString) {
+    if !new_entry.is_empty() {
+        KILL_LIST.lock().unwrap().push_front(new_entry);
+    }
+}
+
+fn kill_replace_ffi(old_entry: &CxxWString, new_entry: &CxxWString) {
+    kill_replace(old_entry.from_ffi(), new_entry.from_ffi())
+}
+
+/// Replace the specified string in the killring.
+pub fn kill_replace(old_entry: WString, new_entry: WString) {
+    let mut kill_list = KILL_LIST.lock().unwrap();
+    kill_list.retain(|str| str != &old_entry);
+    kill_list.push_front(new_entry);
+}
+
+fn kill_yank_rotate_ffi(mut out_front: Pin<&mut CxxWString>) {
+    out_front.as_mut().clear();
+    out_front
+        .as_mut()
+        .push_chars(kill_yank_rotate().as_char_slice());
+}
+
+/// Rotate the killring.
+pub fn kill_yank_rotate() -> WString {
+    let mut kill_list = KILL_LIST.lock().unwrap();
+    kill_list.rotate_left(1);
+    kill_list.front().cloned().unwrap_or_default()
+}
+
+fn kill_yank_ffi(mut out_front: Pin<&mut CxxWString>) {
+    out_front.as_mut().clear();
+    out_front.as_mut().push_chars(kill_yank().as_char_slice());
+}
+
+/// Paste from the killring.
+pub fn kill_yank() -> WString {
+    let kill_list = KILL_LIST.lock().unwrap();
+    kill_list.front().cloned().unwrap_or_default()
+}
+
+pub fn kill_entries_len() -> usize {
+    KILL_LIST.lock().unwrap().len()
+}
+
+fn kill_entries_ffi(mut out_entries: Pin<&mut CxxVector<CxxWString>>) -> bool {
```

Comment:
This is also fine as written but see `env_ffi.rs as_list_ffi()` for a more direct way to do this.

Reply:
Thanks for mentioning. I've changed it to use `wcstring_list_ffi_t`.